src/MongoDB.Driver/TransactionExecutor.cs

```diff
@@ -0,0 +1,308 @@
+            TransactionOptions transactionOptions, 
+            CancellationToken cancellationToken)
+        {
+            TResult result;
+
+            while (true)
+            {
+                clientSession.StartTransaction(transactionOptions);
+
+                var transactionStepResult = WithTransactionCallbackProcessing(clientSession, callback, startTime, dateTimeProvider, cancellationToken);
+                if (transactionStepResult.TransactionShouldBeRestarted)
+                {
+                    continue;
+                }
+
+                result = transactionStepResult.Result;
+
+                if (!clientSession.IsInTransaction)
+                {
+                    return result; // Assume callback intentionally ended the transaction.
+                }
+
+                var transactionHasBeenConfirmed = WithTransactionCommitProcessingWithRetry(clientSession, startTime, dateTimeProvider, cancellationToken);
+                if (transactionHasBeenConfirmed)
+                {
+                    // the transaction has been confirmed.
+                    break;
+                }
+
+                cancellationToken.ThrowIfCancellationRequested();
+            }
+
+            return result;
+        }
+
+        internal static async Task<TResult> ExecuteWithRetriesAsync<TResult>(
+            IClientSessionHandle clientSession,
+            Func<IClientSessionHandle, CancellationToken, Task<TResult>> callbackAsync,
+            DateTime startTime,
+            IDateTimeProvider dateTimeProvider,
+            TransactionOptions transactionOptions,
+            CancellationToken cancellationToken)
+        {
+            TResult result;
+
+            while (true)
+            {
+                clientSession.StartTransaction(transactionOptions);
+
+                var transactionStepResult = await WithTransactionCallbackProcessingAsync(clientSession, callbackAsync, startTime, dateTimeProvider, cancellationToken).ConfigureAwait(false);
+                if (transactionStepResult.TransactionShouldBeRestarted)
+                {
+                    continue;
+                }
+
+                result = transactionStepResult.Result;
+
+                if (!clientSession.IsInTransaction)
+                {
+                    return result; // Assume callback intentionally ended the transaction.
+                }
+
+                var transactionHasBeenConfirmed = await WithTransactionCommitProcessingWithRetryAsync(clientSession, startTime, dateTimeProvider, cancellationToken).ConfigureAwait(false);
+                if (!transactionHasBeenConfirmed)
+                {
+                    // the transaction has been confirmed.
+                    break;
+                }
+
+                cancellationToken.ThrowIfCancellationRequested();
+            }
+
+            return result;
+        }
+
+        private static T GetElementOrDefault<T>(BsonDocument document, string name) where T : BsonValue
```

Comment:
This helper method is not needed. There are existing methods on BsonDocument that accomplish the same thing.

See comments below.

Reply:
Done